[Submission] Liu, Kong, Liu (#14)
mdx-workshop/mdx-submissions21: paper.md

---
title: 'CWS-PResUNet: Music Source Separation with Channel-wise Subband Phase-aware ResUNet'
tags:
  - music
  - separation
  - channel-wise subband
  - resunet
authors:
  - name: Haohe Liu # note this makes a footnote saying 'co-first author'
    orcid: 0000-0003-1036-7888
    affiliation: "1, 2" # (Multiple affiliations must be quoted)
  - name: Qiuqiang Kong # note this makes a footnote saying 'co-first author'
    affiliation: 1
  - name: Jiafeng Liu
    affiliation: 1
affiliations:
 - name: Sound, Audio, and Music Intelligence (SAMI) Group, ByteDance
   index: 1
 - name: The Ohio State University
   index: 2
date: 10 Oct 2021
bibliography: paper.bib
# arxiv-doi: 10.21105/joss.01667
---

# Abstract  

> Music source separation (MSS) shows active progress with deep learning models in recent years. Many MSS models perform separations on spectrogram by estimating bounded ratio masks and reusing the phases of the mixture. When using convolutional neural networks (CNN), weights are usually shared within spectrogram during convolution regardless of the different patterns between frequency bands. In this study, we propose a new MSS model, channel-wise subband phase-aware ResUNet (CWS-PResUNet), to decompose signals into subbands and estimate the source unbound complex ideal ratio mask (cIRM) for MSS. CWS-PResUNet utilizes channel-wise subband (CWS) feature to limit unnecessary global weights sharing on the spectrogram and reduce computational resource consumptions. The saved computational cost and memory can in turn allow for a larger architecture. On the MUSDB18HQ test set, we propose a 276-layer CWS-PResUNet and achieve state-of-the-art (SoTA) performance on `vocals` with an 8.92 signal-to-distortion ratio (SDR) score. By combining CWS-PResUNet and Demucs, our ByteMSS system ranks the 2nd on `vocals` score and 5th on average score in the 2021 ISMIR Music Demixing (MDX) Challenge limited training data track (leaderboard A). Our code and pre-trained models are publicly available^[Open sourced at: https://github.com/haoheliu/2021-ISMIR-MSS-Challenge-CWS-PResUNet].

Reviewer: ```suggestion
Music source separation (MSS) shows active progress with deep learning models in recent years. Many MSS models perform separations on spectrograms by estimating bounded ratio masks and reusing the phases of the mixture. When using convolutional neural networks (CNN), weights are usually shared within a spectrogram during convolution regardless of the different patterns between frequency bands. In this study, we propose a new MSS model, channel-wise subband phase-aware ResUNet (CWS-PResUNet), to decompose signals into subbands and estimate for each source an unbounded complex ideal ratio mask (cIRM) for MSS. CWS-PResUNet utilizes a channel-wise subband (CWS) feature to limit unnecessary global weights sharing on the spectrogram and reduce computational resource consumptions. The saved computational cost and memory can in turn allow for a larger architecture. On the MUSDB18HQ test set, we propose a 276-layer CWS-PResUNet and achieve state-of-the-art (SoTA) performance on `vocals` with an 8.92 signal-to-distortion ratio (SDR) score. By combining CWS-PResUNet and Demucs, our ByteMSS system ranks the 2nd on `vocals` score and 5th on average score in the 2021 ISMIR Music Demixing (MDX) Challenge limited training data track (leaderboard A). Our code and pre-trained models are publicly available^[Open sourced at: https://github.com/haoheliu/2021-ISMIR-MSS-Challenge-CWS-PResUNet].
```

# Introductions and Related Works

> Music source separation aims at decomposing a music mixture into several soundtracks, such as `Vocals`, `Bass`, `Drums`, and `Other` tracks. It is closely related to topics like music transcription, remixing, and retrieval. Based on deep learning models, most of the early studies [@jansson2017singing;@takahashi2018mmdenselstm] perform separations in the frequency domain by estimating the ideal ratio masks (IRM) of the magnitude spectrogram and reusing the phase of the mixture. Later, time-domain models [@defossez2019demucs] start to demonstrate SoTA performance using direct waveform modeling, which does not involve transformations like short-time fourier transform (STFT). In this case, phase information can be implicitly estimated and models will not be restricted with the fixed time-frequency resolution. To enhance the MSS performance, @liu2020voice chooses to employ self-attension mechanism and Dense-UNet architecture. @choi2019investigating compares the performance of several types of UNet built with different intermediate blocks. To alleviate the computational cost, @kadandale2020multi designs a multi-task model to replace source-dedicated models. Also, @liu2020channel proposes to use the channel-wise subband feature to reduce resource consumptions and improve separation performance. Recently, @kong2021decoupling conducts experiment on the MSS system theoretical upper bound, which proves the limitation of IRMs and the importance of phase estimation. 

Reviewer: I would use past tense here to summarize the previous literature - e.g., `Later, time-domain models [@defossez2019demucs] started to demonstrate ...`.

Reviewer: ```suggestion
To enhance the MSS performance, @liu2020voice chose to employ a self-attention mechanism and Dense-UNet architecture.
```

Reviewer: ```suggestion
Recently, @kong2021decoupling conducted experiments on the theoretical upper bound for MSS, which proved the limitation of IRMs and the importance of phase estimation.
```

In the next section, we will introduce the detailed architecture of CWS-PResUNet as well as ByteMSS, the system we submitted for the MDX Challenge [@mitsufuji2021music]. 

![Overview of the CWS-PResUNet and a comparison between using magnitude spectrogram and channel-wise subband spectrogram as the input feature.^[We use mono signal for simple illustration.]](graphs/main.png){ width=100% }

# Method

> CWS-PResUNet is a ResUNet [@zhang2018road;@liu2021voicefixer] based model integrating the CWS feature [@liu2020channel] and the cIRM estimation strategies described in @kong2021decoupling. The overall pipeline is summarized in Figure 1a. For a stereo mixture signal $x \in \mathbb{R}^{2\times L}$, where $L$ stands for signal length, we first utilize pre-defined analysis filters ${h}^{(j)},j=1,2,3,4$ to perform subband decompositions:

Reviewer: > ... we first utilize pre-defined analysis filters ${h}^{(j)},j=1,2,3,4$ to perform subband decompositions

It would be good if you could specify the analysis/synthesis filters that you used for your filter bank. Which set of filters did you use? Are they defining a perfect reconstruction filter bank?

$$
x^{\prime}_{8\times \frac{L}{4}} = [\text{DS}_4({x_{2\times 1 \times L}}*{h}^{(j)}_{1\times 64})]_{j=1,2,3,4},
$$
where $\text{DS}_{4}(\cdot)$, $*$, and $[\cdot]$ denote the downsampling by 4, convolution, and stacking operators, respectively. Then we calculate the STFT of the downsampled subband signals $x^{\prime}$ to obtain their magnitude spectrograms $|X^{\prime}|_{8\times T \times \frac{F}{4}}$. 

![The architecture of Phase-aware ResUNet](graphs/arc.png){ width=100% }

As is shown in Figure 2, the phase-aware ResUNet is a symmetric architecture containing a down-sampling and an up-sampling path with skip-connections between the same level. It accepts $|X^{\prime}|$ as input and estimates four tensors with the same shape: mask estimation $\hat{M}$, phase variation $\hat{P}_{r}$, $\hat{P}_{i}$, and direct magnitude prediction $\hat{Q}$. The complex spectrogram can be reconstructed with the following equation:
$$

> \hat{S}^{\prime} = \text{relu}(|X^{\prime}|\odot \text{sigmoid}(\hat{M})+\hat{Q})e^{j(\angle X^{\prime} +\angle \hat{\theta})},

Reviewer 2: maybe \exp would be more readable ?

$$
in which $cos\angle \hat{\theta}=\hat{P}_{r}/(\sqrt{\hat{P}_{r}^2+\hat{P}_{i}^2})$ and $sin\angle \hat{\theta}=\hat{P}_{i}/(\sqrt{\hat{P}_{r}^2+\hat{P}_{i}^2})$. We pass the mask estimation $\hat{M}$ through a sigmoid function to obtain a mask with values between 0 and 1. Then by estimating $\hat{Q}$ and $\hat{\theta}$, models can avoid estimating mask with only bounded values and using mixture phase to calculate the unbounded cIRM. We use relu activation to ensure the positve magnitude value. Finally, after the inverse STFT, we perform subband reconstructions to obtain the source estimation $\hat{s}$:
$$
\hat{s}_{2\times L} = \sum_{j=1}^{4}(\text{US}_4(\hat{s}^{\prime}_{2\times 4\times \frac{L}{4}})*g^{(j)}_{4\times 64}),
$$
where $g^{(j)}, j=1,2,3,4$ are the pre-defined synthesis filters and $\text{US}_4(\cdot)$ is the zero-insertion upsampling function.

> As is illustrated in Figure 1b, the CWS feature can make the CNN feature-maps smaller and save computational resources. Also, models become more efficient by enlarging receptive fields and diverging subband information into different channels. Our model for `vocals` is optimized by calculating L1 loss between $\hat{s}$ and its target source $s$. Despite we also use a model dedicated to separating the `other` track, we notice estimating and optimizing four sources together can result in a 0.2 SDR [@vincent2006performance] gain on `other`. So, we calculate both L1 loss and energy-conservation loss across four sources to optimize the model for `other`. Our CWS-PResUNet models for `bass` and `drums` reported in the next section employ the same setup as the model for `other`.

Reviewer 2: * "Despite we": although we
* could be describe your "energy conservation loss" exactly ?

> As is illustrated in Figure 1b, the CWS feature can make the CNN feature-maps smaller and save computational resources. Also, models become more efficient by enlarging receptive fields and diverging subband information into different channels. Our model for `vocals` is optimized by calculating L1 loss between $\hat{s}$ and its target source $s$. Although we also use a model dedicated to separating the `other` track, we notice estimating and optimizing four sources together in one model can result in a 0.2 SDR [@vincent2006performance] gain on `other`. In this case, we not only use L1 loss on the waveform, but also employ energy-conservation loss, which calculate the L1 loss between mixture and the sum of four source estimations. Our CWS-PResUNet models for `bass` and `drums` reported in the next section employ the same setup as the model for `other`.

Reviewer: > As is illustrated in Figure 1b, the CWS feature can make the CNN feature-maps smaller and save computational resources.

Is the computational complexity really reduced? Using the subband decomposition, we decrease the shape of the tensor in the temporal direction by a factor 4 but increase the number of feature maps by 4. Do you increase the number of output feature maps for all convolutions by 4 in the model? If yes, then the computational complexity actually increased as we now need to compute 4*4 = 16 times more convolutions of feature maps that have only a factor 4 smaller size in one dimension.

Author: Thanks for pointing out that. I'll do further clearance on this point in our paper.

First of all, only the channel of the first input conv2d and the last output conv2d is increased by 4. Any other model settings are unchanged.
Secondly, we decrease the frequency dimension by 4. Since the channel setting of most of the con2d (except the very first and the very last) is not changed, the complexity of the model will decrease by 4.

Reviewer: ok, I understood - then everything is fine 👍

<!-- Moreover, because bounded mask and mixture phase can limit the theoretical upper bound of the MSS system [@kong2021decoupling], we estimate unbounded mask and phase variations in each subband to compute the unbounded cIRM.  -->

> In our ByteMSS system, we set up Demucs [@defossez2019demucs] to separate `bass` and `drums` tracks because it performs better than CWS-PResUNet on these two sources. Demucs is a time-domain MSS model. In our study, we adopted the open-sourced pre-trained Demucs^[https://github.com/facebookresearch/demucs] and do not apply the shift trick because it will slow down the inference speed. Also, we utilize a 276-layer CWS-PResUNet to separate the `vocals` track and a 166-layer CWS-PResUNet for the `other` track. We set the latter with fewer layers for faster inference speed concerns. 

Reviewer: > We set the latter with fewer layers for faster inference speed concerns.

Did you see also a difference in generalization performance? For `other`, I can imagine that overfitting is more of a problem and, hence, it is also better to use a smaller model for this instruments.

# Experiments

Our models are optimized using the training subset of MUSDB18HQ [@rafii2019musdb18]. We calculate the STFT of the downsampled 11.05 kHz subband signals with a window length of 512 and a window shift of 110. We use Adam optimizer with an initial learning rate of 0.001 and exponential decay. CWS-PResUNet takes approximately four days to train on a Tesla V100 GPU. During inference, we utilize a 10-second long `boxcar` windowing function [@schuster2008influence] with no overlapping to segment the signal. For evaluation, we report the SDR on the MUSDB18HQ test set with the open-sourced *museval* tool [@SiSEC18]. 

> The subband analysis and synthesis operations usually cannot achieve perfect reconstruction. To assess the errors introduced by subband operations, we decompose the test set `vocals` tracks into 2,4, and 8 subbands and reconstruct them back for evaluations. As is presented in Table 1, in all cases subband reconstructions achieve high performance with only neglectable errors.

Reviewer 2: reconstruct them back to evaluate the reconstruction error of the filterbank

> <!-- | Subband numbers |   2   |   4  |   8  |

Reviewer: A SDR of 79.9dB looks still like a perfect reconstruction as it is larger than the error that we would expect if we use float precision (for float precision, the machine precision is `eps = 1.19e-07`. With which precision did you compute the construction error (float, double)?

Author: Yes, that's a good point. I do the calculation using 32  bits float number. But strictly speaking, 79.9dB does not mean perfect reconstruction. I try adding 1e-9 to some numbers to evaluate SDR and get a value around 180dB. So the error introduced by subband operation results in this gap between 180dB and 79.9dB.
